### ISSUE TITLE
Where can I find the DSA W weight training code? The one I used with AI generated weights is not converging.

### ISSUE THREAD

**Issue by zhangzhiyuan303** [deepseek-ai/DeepSeek-V3.2-Exp#47]
I want to use this DSA method in another project, but I found that it requires training a network with two Linnear and one weight (w). I haven't been able to implement it successfully, and I haven't found any solutions online. I hope you can give me some ideas.

**The main problem is non-convergence.**  lora :d512--d128
class SimpleSparseAttention(nn.Module):
    def __init__(self, config):
        super().__init__()
        self.config = config
        
        self.q_proj = nn.Linear(128, 32)
        self.k_proj = nn.Linear(128, 32)
        
        nn.init.normal_(self.q_proj.weight, std=0.01)
        nn.init.normal_(self.k_proj.weight, std=0.01)
        nn.init.zeros_(self.q_proj.bias)
        nn.init.zeros_(self.k_proj.bias)
def forward(self, q, k):

        q_proj = self.q_proj(q)
        k_proj = self.k_proj(k)        

        attn_scores = torch.matmul(q_proj, k_proj.transpose(-1, -2))
        attn_scores = attn_scores / np.sqrt(self.config.d_model)

        attn_scores = F.relu(attn_scores)
        

        attn_scores_max = torch.max(attn_scores, dim=-1, keepdim=True)[0]
        attn_scores = torch.exp(attn_scores - attn_scores_max)
        attn_scores = attn_scores / (torch.sum(attn_scores, dim=-1, keepdim=True) + self.config.eps)
        

        values, indices = torch.topk(attn_scores, self.config.top_k, dim=-1)
        
         sparse_attn = torch.zeros_like(attn_scores)
        sparse_attn.scatter_(-1, indices, values)
        
        sparse_attn = sparse_attn / (torch.sum(sparse_attn, dim=-1, keepdim=True) + self.config.eps)
        
        return sparse_attn, indices, attn_scores
